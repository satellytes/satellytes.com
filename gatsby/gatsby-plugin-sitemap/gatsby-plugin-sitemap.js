const TRANSLATION_KEY_REGEX = /^\/(de)(\/.*)$/;
const DEFAULT_LANGUAGE = 'en';

/**
 * Extract the language key from the given path.
 * return the language key or if none is found it returns the default
 * language'.
 *
 * Example:
 *
 * `/de/simple-path` -> de
 * `/simple-path` -> en
 *
 * This will also mean that unsupported languages are ignored and treated as english
 * `/es/simple-path` -> en
 */
const extractLangKey = (path) => {
  const match = TRANSLATION_KEY_REGEX.exec(path);
  if (match) {
    const [_, languageKeyMatched] = match;
    return languageKeyMatched;
  }

  return DEFAULT_LANGUAGE;
};

/**
 * Given a url try to remove the first path segment if it matches a translated
 * url which we check through a static regex (which currently only involves the check for `de`)
 */
const normalizePath = (path) => {
  const match = TRANSLATION_KEY_REGEX.exec(path);
  if (match) {
    const [, , remainingPath] = match;
    return remainingPath;
  }

  return path;
};

/**
 * Given a normalized path where any valid language key is stripped
 * we can derive the translated url.
 * Example: Normalized path is `/my-page` and we can now create two urls from here:
 * `createTranslatedUrl('/my-page', 'de')`
 * Result: `/de/my-page`
 *
 * `createTranslatedUrl('/my-page', 'en')`
 * Result:  `/my-page` (unchanged, because it's our default language)
 */
const createTranslatedUrl = (normalizedPath, language) => {
  if (language === DEFAULT_LANGUAGE) {
    return normalizedPath;
  }
  return `/${language}${normalizedPath}`;
};

/**
 * Given a set of pages nodes which only need carry a `path` attribute,
 * we traverse all of the pages and group them by their `normalized` path
 * which is the path stripped of any known language key. That way we have
 * clusters of translated pages.
 *
 * The map looks as follows for three given paths `/some-page`, `/de/some-page` and `/unrelated-page`.
 * ```
 *  Map(2) {
 *     '/some-page' => Set(2) { 'en', 'de' },
 *    '/unrelated-page' => Set(1) { 'en' }
 *  }
 * ```
 */
const createPageTranslationLookup = (paths) => {
  const map = new Map();

  for (path of paths) {
    const normalizedPageId = normalizePath(path);

    if (!map.has(normalizedPageId)) {
      map.set(normalizedPageId, new Set());
    }

    const languageKey = extractLangKey(path);
    const translationSet = map.get(normalizedPageId);
    translationSet.add(languageKey);
  }
  return map;
};

/**
 * Given a normalizedPath and a set of languages
 * we can construct a list of links that tell about their `lang` and the actual `path`
 * to the translation
 */
const createTranslationLinks = (normalizedPath, languages) => {
  let result = [];
  for (const lang of languages.values()) {
    result.push({
      lang,
      path: createTranslatedUrl(normalizedPath, lang),
    });
  }

  return result;
};

/**
 * Gatsby will invoke the following three function in the order `resolvePages`, `filteredPages` & `serialize`
 * The actual invocation looks like this
 * ```
 * const resolvedPages = resolvePages(PAGE_QUERY_RESULT)
 * const filteredPages = resolvedPages.filter(
 *   page => !excludes.some(excludedRoute => filterPages(page, excludedRoute))
 * )
 * const sitemapResult = filteredPages.map(page => serialize(page));
 * ```
 */
function gatsbyPluginSitemap({ allSitePage: { nodes: allPages } }) {
  const paths = allPages.map(({ path }) => path);
  const translationLookup = createPageTranslationLookup(paths);

  return allPages.map(({ path, pageContext }) => {
    const normalizedPath = normalizePath(path);
    const languageKey = extractLangKey(path);

    const translations = translationLookup.get(normalizedPath);
    const links = createTranslationLinks(normalizedPath, translations);

    return {
      lang: languageKey,
      path,
      links,
      publicationDate: pageContext?.publicationDate,
    };
  });
}

function serialize(sitemapItem) {
  const HIGH_PRIO =
    sitemapItem.path.includes('/blog') ||
    sitemapItem.path.includes('/career') ||
    sitemapItem.path === '/' ||
    sitemapItem.path === '/de';

  // remap our links from {lang, path} to {lang, url}
  const alternateLinks = sitemapItem.links.map((item) => {
    return {
      lang: item.lang,
      url: item.path, // the plugin will prefix the siteUrl after the serialization
    };
  });

  return {
    url: sitemapItem.path,

    // there is a lot of fuzz about changefreq and priority, some even say it
    // doesn't have any effect at all. so let's use a very simple definition
    changefreq: 'daily',
    priority: HIGH_PRIO ? 1 : 0.8,

    // publicationDate is optional and won't be rendered if not present
    lastmod: sitemapItem.publicationDate,
    links: alternateLinks,
  };
}

function filterPages({ path }, excludes) {
  const [EXCLUDE, KEEP] = [true, false];

  if (excludes === path) {
    return EXCLUDE;
  }

  if (!path.endsWith('/')) {
    console.warn(
      `Path '${path}' does not end with a slash! For SEO reasons all paths should end with a slash`,
    );
  }

  return KEEP;
}

module.exports = {
  resolvePages: gatsbyPluginSitemap,
  serialize,
  filterPages,
};
