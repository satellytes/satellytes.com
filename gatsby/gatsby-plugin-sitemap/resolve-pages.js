const TRANSLATION_KEY_REGEX = /^\/(de|es)(\/.*)$/;
const DEFAULT_LANGUAGE = 'en';
const { groupBy, uniq } = require('lodash');

/**
 * extract the language key from the given path.
 * return the language key or if none is found it returns the default
 * language'
 */
const extractLangKey = (path) => {
  const match = TRANSLATION_KEY_REGEX.exec(path);
  if (match) {
    const [_, languageKeyMatched] = match;
    return languageKeyMatched;
  }

  return DEFAULT_LANGUAGE;
};
/**
 * if a language key is present remove it to get the expected path for the default
 * language to group translated pages based on this value later.
 */
const normalizePath = (path) => {
  const match = TRANSLATION_KEY_REGEX.exec(path);
  if (match) {
    const [, , remainingPath] = match;
    return remainingPath;
  }

  return path;
};

/**
 * transform all raw page nodes into a format we can use in the serialization process
 * to create the sitemap entries
 */

function createTranslatedUrl(masterId, language) {
  if (language === DEFAULT_LANGUAGE) {
    return masterId;
  }
  return `/${language}${masterId}`;
}

function createPageTranslationLookup(pages) {
  const map = new Map();

  for ({ path } of pages) {
    const masterID = normalizePath(path);

    if (!map.has(masterID)) {
      map.set(masterID, new Set());
    }

    const languageKey = extractLangKey(path);
    const translationSet = map.get(masterID);
    translationSet.add(languageKey);
  }

  return map;
}

function createTranslationLinks(masterID, languages) {
  let result = [];
  for (const language of languages.values()) {
    result.push({
      language,
      path: createTranslatedUrl(masterID, language),
    });
  }

  return result;
}

function resolvePages({ allSitePage: { nodes: allPages } }) {
  const translationLookup = createPageTranslationLookup(allPages);

  return allPages.map(({ path }) => {
    const masterID = normalizePath(path);
    const translations = translationLookup.get(masterID);
    const links = createTranslationLinks(masterID, translations);

    return {
      lang: extractLangKey(path),
      path,
      links,
    };
  });
}

function serialize(sitemapItem) {
  console.log('*** serialize', sitemapItem);

  // { lang: 'de', url: isGermanPath ? path : translation },
  return {
    url: sitemapItem.path,
    changefreq: 'daily',
    priority: 0.7,
    links: sitemapItem.links.map((item) => {
      return {
        lang: item.lang,
        url: item.path,
      };
    }),
  };
}

function filterPages(page, excludes) {
  console.log('*** filterPages', page);
  if (excludes === page.path) {
    return true; // excludes page
  }
  if (!page.path.endsWith('/')) {
    console.warn(
      'Path of the page does not end with a slash! For SEO reasons all paths should end with a slash:',
      page.path,
    );
  }
  return false;
}

module.exports = {
  resolvePages,
  serialize,
  filterPages,
};
