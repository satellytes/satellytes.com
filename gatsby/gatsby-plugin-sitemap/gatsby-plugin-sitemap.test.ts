import { resolvePages, serialize, filterPages } from './gatsby-plugin-sitemap';

const EXAMPLES_NODES = {
  allSitePage: {
    nodes: [
      { path: '/some-page/' },
      { path: '/de/some-page/' },
      { path: '/de/only-german/' },
      { path: '/only-default/' },
      { path: '/ignore-me/' },
    ],
  },
};

const createPagesNodes = (nodes) => {
  return { allSitePage: { nodes } };
};

describe('Gatsby Plugin Sitemap', () => {
  test('works', () => {
    const allPages = resolvePages(EXAMPLES_NODES);
    const excludes = ['/ignore-me/'];
    const filteredPages = allPages.filter(
      (page) =>
        !excludes.some((excludedRoute) => filterPages(page, excludedRoute)),
    );
    const result = filteredPages.map((page) => serialize(page));
    expect(result).toMatchSnapshot();
  });

  describe('filterPages', () => {
    test('excludes a page', () => {
      const result = filterPages({ path: 'exclude-me/' }, 'exclude-me/');
      expect(result).toBe(true);
    });
    test('accepts a page', () => {
      const result = filterPages({ path: 'include-me/' }, 'exclude-me/');
      expect(result).toBe(false);
    });
  });

  describe('resolvePages', () => {
    test("a page without a language in the path receives the 'en' language (default)", () => {
      const nodes = createPagesNodes([{ path: '/some-page/' }]);
      const [result] = resolvePages(nodes);
      expect(result.lang).toBe('en');
    });

    test('determine the correct language from the path', () => {
      const nodes = createPagesNodes([{ path: '/de/some-page' }]);
      const [result] = resolvePages(nodes);
      expect(result.lang).toBe('de');
    });

    /**
     * well that could be handled better but we don't want to over engineer things for now
     */
    test('unknown languages (eg. `es`) are ignored and treated as a default url', () => {
      const nodes = createPagesNodes([{ path: '/es/some-page' }]);
      const [result] = resolvePages(nodes);
      expect(result.lang).toBe('en');
    });
    test("two related pages link each other's translations", () => {
      const nodes = createPagesNodes([
        { path: '/some-page/' },
        { path: '/de/some-page/' },
        { path: '/unrelated-page/' },
      ]);
      const [page1, page2, _] = resolvePages(nodes);
      expect(page1).toEqual(
        expect.objectContaining({ lang: 'en', path: '/some-page/' }),
      );
      expect(page2).toEqual(
        expect.objectContaining({ lang: 'de', path: '/de/some-page/' }),
      );
      expect(page1.links).toEqual(page2.links);
    });
  });

  describe('serialize', () => {
    test('creates a correct sitemap entry', () => {
      const result = serialize({
        path: '/some-page/',
        language: 'en',
        links: [
          {
            lang: 'en',
            path: 'abc/',
          },
          {
            lang: 'de',
            path: 'de/abc/',
          },
        ],
      });

      expect(result).toEqual({
        url: '/some-page/',
        priority: 0.8,
        lastmod: undefined,
        changefreq: 'daily',
        links: [
          {
            lang: 'en',
            url: 'abc/',
          },
          {
            lang: 'de',
            url: 'de/abc/',
          },
        ],
      });
    });
  });
});
