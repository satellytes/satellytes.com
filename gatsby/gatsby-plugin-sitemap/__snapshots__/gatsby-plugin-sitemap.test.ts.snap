// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gatsby Plugin Sitemap works 1`] = `
Array [
  Object {
    "changefreq": "daily",
    "lastmod": undefined,
    "links": Array [
      Object {
        "lang": "en",
        "url": "/some-page/",
      },
      Object {
        "lang": "de",
        "url": "/de/some-page/",
      },
    ],
    "priority": 0.8,
    "url": "/some-page/",
  },
  Object {
    "changefreq": "daily",
    "lastmod": undefined,
    "links": Array [
      Object {
        "lang": "en",
        "url": "/some-page/",
      },
      Object {
        "lang": "de",
        "url": "/de/some-page/",
      },
    ],
    "priority": 0.8,
    "url": "/de/some-page/",
  },
  Object {
    "changefreq": "daily",
    "lastmod": undefined,
    "links": Array [
      Object {
        "lang": "de",
        "url": "/de/only-german/",
      },
    ],
    "priority": 0.8,
    "url": "/de/only-german/",
  },
  Object {
    "changefreq": "daily",
    "lastmod": undefined,
    "links": Array [
      Object {
        "lang": "en",
        "url": "/only-default/",
      },
    ],
    "priority": 0.8,
    "url": "/only-default/",
  },
]
`;
